server-common/src/main/java/org/apache/kafka/server/common/MetadataVersion.java

```diff
@@ -0,0 +1,290 @@
+
+    // Introduced LeaderAndIsrRequest V1, UpdateMetadataRequest V4 and FetchRequest V6 via KIP-112
+    IBP_1_0_IV0(-1, "1.0", "IV0"),
+
+    // Introduced DeleteGroupsRequest V0 via KIP-229, plus KIP-227 incremental fetch requests,
+    // and KafkaStorageException for fetch requests.
+    IBP_1_1_IV0(-1, "1.1", "IV0"),
+
+    // Introduced OffsetsForLeaderEpochRequest V1 via KIP-279 (Fix log divergence between leader and follower after fast leader fail over)
+    IBP_2_0_IV0(-1, "2.0", "IV0"),
+
+    // Several request versions were bumped due to KIP-219 (Improve quota communication)
+    IBP_2_0_IV1(-1, "2.0", "IV1"),
+
+    // Introduced new schemas for group offset (v2) and group metadata (v2) (KIP-211)
+    IBP_2_1_IV0(-1, "2.1", "IV0"),
+
+    // New Fetch, OffsetsForLeaderEpoch, and ListOffsets schemas (KIP-320)
+    IBP_2_1_IV1(-1, "2.1", "IV1"),
+
+    // Support ZStandard Compression Codec (KIP-110)
+    IBP_2_1_IV2(-1, "2.1", "IV2"),
+
+    // Introduced broker generation (KIP-380), and
+    // LeaderAdnIsrRequest V2, UpdateMetadataRequest V5, StopReplicaRequest V1
+    IBP_2_2_IV0(-1, "2.2", "IV0"),
+
+    // New error code for ListOffsets when a new leader is lagging behind former HW (KIP-207)
+    IBP_2_2_IV1(-1, "2.2", "IV1"),
+
+    // Introduced static membership.
+    IBP_2_3_IV0(-1, "2.3", "IV0"),
+
+    // Add rack_id to FetchRequest, preferred_read_replica to FetchResponse, and replica_id to OffsetsForLeaderRequest
+    IBP_2_3_IV1(-1, "2.3", "IV1"),
+
+    // Add adding_replicas and removing_replicas fields to LeaderAndIsrRequest
+    IBP_2_4_IV0(-1, "2.4", "IV0"),
+
+    // Flexible version support in inter-broker APIs
+    IBP_2_4_IV1(-1, "2.4", "IV1"),
+
+    // No new APIs, equivalent to 2.4-IV1
+    IBP_2_5_IV0(-1, "2.5", "IV0"),
+
+    // Introduced StopReplicaRequest V3 containing the leader epoch for each partition (KIP-570)
+    IBP_2_6_IV0(-1, "2.6", "IV0"),
+
+    // Introduced feature versioning support (KIP-584)
+    IBP_2_7_IV0(-1, "2.7", "IV0"),
+
+    // Bup Fetch protocol for Raft protocol (KIP-595)
+    IBP_2_7_IV1(-1, "2.7", "IV1"),
+
+    // Introduced AlterPartition (KIP-497)
+    IBP_2_7_IV2(-1, "2.7", "IV2"),
+
+    // Flexible versioning on ListOffsets, WriteTxnMarkers and OffsetsForLeaderEpoch. Also adds topic IDs (KIP-516)
+    IBP_2_8_IV0(-1, "2.8", "IV0"),
+
+    // Introduced topic IDs to LeaderAndIsr and UpdateMetadata requests/responses (KIP-516)
+    IBP_2_8_IV1(-1, "2.8", "IV1"),
+
+    // Introduce AllocateProducerIds (KIP-730)
+    IBP_3_0_IV0(1, "3.0", "IV0"),
+
+    // Introduce ListOffsets V7 which supports listing offsets by max timestamp (KIP-734)
+    // Assume message format version is 3.0 (KIP-724)
+    IBP_3_0_IV1(2, "3.0", "IV1"),
+
+    // Adds topic IDs to Fetch requests/responses (KIP-516)
+    IBP_3_1_IV0(3, "3.1", "IV0"),
+
+    // Support for leader recovery for unclean leader election (KIP-704)
+    IBP_3_2_IV0(4, "3.2", "IV0"),
+
+    IBP_3_3_IV0(5, "3.3", "IV0");
```

Comment:
Thinking about it again, do we actually want to introduce this now or wait until we have a reason to do it? Otherwise, we are burning this version with no reason now.

Reply:
Sure, perhaps this doesn't really belong in the refactor